hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerRatisServer.java

```diff
@@ -156,24 +159,48 @@ private OMResponse processReply(OMRequest omRequest, RaftClientReply reply)
     // NotLeader exception is thrown only when the raft server to which the
     // request is submitted is not the leader. This can happen first time
     // when client is submitting request to OM.
-    NotLeaderException notLeaderException = reply.getNotLeaderException();
-    if (notLeaderException != null) {
-      throw new ServiceException(notLeaderException);
-    }
-    StateMachineException stateMachineException =
-        reply.getStateMachineException();
-    if (stateMachineException != null) {
-      OMResponse.Builder omResponse = OMResponse.newBuilder();
-      omResponse.setCmdType(omRequest.getCmdType());
-      omResponse.setSuccess(false);
-      omResponse.setMessage(stateMachineException.getCause().getMessage());
-      omResponse.setStatus(parseErrorStatus(
-          stateMachineException.getCause().getMessage()));
-      if (LOG.isDebugEnabled()) {
-        LOG.debug("Error while executing ratis request. " +
-            "stateMachineException: ", stateMachineException);
+
+    if (!reply.isSuccess()) {
+      NotLeaderException notLeaderException = reply.getNotLeaderException();
+      if (notLeaderException != null) {
+        RaftPeerId suggestedLeader =
+            notLeaderException.getSuggestedLeader() != null ?
+                notLeaderException.getSuggestedLeader().getId() : null;
+        org.apache.hadoop.ozone.om.exceptions.NotLeaderException
+            notLeaderException1;
+        if (suggestedLeader != null) {
+          notLeaderException1 = new org.apache.hadoop.ozone.om.exceptions
+              .NotLeaderException(getRaftPeerId(), suggestedLeader);
+        } else {
+          notLeaderException1 =
+              new org.apache.hadoop.ozone.om.exceptions
+                  .NotLeaderException(getRaftPeerId());
+        }
+        throw new ServiceException(notLeaderException1);
+      }
+
```

Comment:
Do you think it would be good idea to rename org.apache.hadoop.ozone.om.exceptions.NotLeaderException to OMNotLeaderException so as to distinguish with Ratis NotLeaderException.
Also, we can add a method in OMNotLeaderException to parse Ratis NotLeaderException and call that method here.

Reply:
On high level look to ratis, NotReplicatedException is thrown for watch requests. As we don't use any watch requests in OM, I think we should be fine, as in our case we should never see this kind of exception. I will explore more if some handling is needed for this.